### PR TITLE
Right click menu in note entry mode

### PR DESCRIPTION
**Pull request by 19joho66**
This PR introduces a right-click menu in note entry mode. This right click context menu provides the note entry toolbar. This makes the useage of the mouse more comfortable and faster, because one can for example change the note duration at the place of the note input an doesn't need to move the mouse up to the toolbar. 

This PR also includes a new menuentry for the right click context menu im note edit mode. This entry enables the switching to note entry mode. Therefore it is now possible to switch between note entry and note edit mode via right click context menues.

### CODE REVIEW

**Comment by MarcSabatella** [comment]
I do like the idea.  Maybe I missed it, but it probably would have been best to discuss it first on IRC, the developers mailing list, or forums to get feedback.  I suspect people will have suggestions on what particular items should be included.  I'm not so sure about the use of an actual toolbar as opposed to a standard menu.

**Comment by 19joho66** [comment]
No, you didn't miss it. I havn't discussed this issue yet, but I'm willing to do so. The toolbar as context menu has the advantage, that the user sees at once where to click to change for example the duration, add a dot or an accidental opposed to reading menutexts first.
With this PR and the one about the shadow notes I try to improve the usability by making it a bit faster and maybe more intuitive. I think it makes it much easier to enter notes, if you just have to right click directly at the spot where you enter them to select the new duration. This saves you from moving the mousepointer away, up to the toolbar and then you have to find the spot again where you last entered a note.

**Comment by ericfont** [comment]
I've tried out your PR, and I like it alot...definatley saves a lot of mouse work, especially for users who are avoiding keyboard.

One suggestion for your PR is that I'm only able to enter "Note Input Mode" if I right click on an empty space between *systems*, but it would be much more easier to access if I could right click anywhere and find the "Note Input Mode" button in menu.  Currently if I right click on a measure or between staffs in a system, I am unable to access "Note Input Mode".

**Comment by MarcSabatella** [comment]
While we're at it, I've often thought that the Style / General and Style /
Text should be available in the generic context menus that appear when no
element is selected (including empty spots within measures).  Not sure why,
but I keep expecting it there even though I've been using MuseScore for 5
years and probabyl know it as well as anyone :-).  Maybe Layout / Page
Settings too.

**Comment by ericfont** [comment]
I would agree there could be a few more things in the right click menu, including Style->{General,Text}.  However, I fear there will never be an agreement as to what exactly should belong there and not belong there.  Do you know how feasabile it would be to allow the context menu to be user-configurable via Preferences or .ini (and note require a re-compile)?

File: mscore/scoreview.cpp [hunk 1]
```diff
@@ -34,6 +34,7 @@
 #include "textcursor.h"
 #include "textpalette.h"
 #include "texttools.h"
+#include <qtoolbar.h>
```
Comment (ericfont) [inline mscore/scoreview.cpp]: I'm able to compile on my linux without this include line.  It must be already include somewhere up in the include hierarcy, so I don't think it should be included here (at it looks out of place since every other include is a "musescorefile.h".

Comment (lasconic) [inline mscore/scoreview.cpp]: Yes. Qt headers are included in all.h which is used for the precompiled header optimisation (on linux and macosx).

**Comment by ZackTheCardshark** [comment]
<blockquote>While we're at it, I've often thought that the Style / General and Style / Text should be available in the generic context menus that appear when no element is selected (including empty spots within measures).</blockquote>

While we're at it, I've often thought the same thing—and similarly that it should be possible to change text styles via the context menu, and also that more elements should link to their section of the General Style settings. But I have to point out that this would <i>require</i> the change to "Edit General Style…" and "Edit Text Styles…" that I recently proposed. Right-clicking and seeing "General…" and "Text…" in the context menu would be as clear as mud.

**Comment by ZackTheCardshark** [comment]
By the way, as to the specific change proposed here, I like it a lot.

**Comment by 19joho66** [comment]
I've introduced this feature in the feature request forum, in order to collect some opinions about this.

**Comment by 19joho66** [comment]
There has been a suggestion for an additional feature in the forum, which i'd like to implement before a merge. 
I will write a comment when the featurre is coded.

**Comment by 19joho66** [comment]
Adde menu entry in all right click context menues in note edit mode, which allows  the user to switch to note entry mode

File: mscore/scoreview.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [inline mscore/scoreview.cpp]: So better remove that line

**Comment by Jojo-Schmitz** [comment]
oops, you've (again) merged rather than rebased.
Please squash into one commit.

The feature request is the one discussed in https://musescore.org/en/node/99536, right?

File: .gitignore [hunk 2]
```diff
@@ -26,3 +26,6 @@ CMakeLists.txt.user.*
 vtest/html
 thirdparty/mupdf-qt
 
+/BuildRelease.bat
+/BuildClean.bat
+/BuildInstall.bat
```
Comment (Jojo-Schmitz) [inline .gitignore]: after these changes these Build*.bat files should not be part of the commit, should they?

File: mscore/scoreview.cpp [hunk 3]
```diff
@@ -695,6 +695,102 @@ ScoreView::ScoreView(QWidget* parent)
       _curLoopOut = new PositionCursor(this);
       _curLoopOut->setType(CursorType::LOOP_OUT);
 
+
+
```
Comment (Jojo-Schmitz) [inline mscore/scoreview.cpp]: here and in other places quite a few added empty lines, seems too much to me

File: mscore/scoreview.cpp [hunk 4]
```diff
@@ -695,6 +695,102 @@ ScoreView::ScoreView(QWidget* parent)
...
+            }
+
+      entryContxtToolbar->addWidget(contxtToolbarWdg);
+      entryContxtMenu = new QMenu(this);
+      
+      QWidgetAction* noteAction = new QWidgetAction(entryContxtMenu);
+      noteAction->setDefaultWidget(entryContxtToolbar);
+      entryContxtMenu->addAction(noteAction);
+
+      for(int i=0;i<actionArray.count();i++)
```
Comment (Jojo-Schmitz) [inline mscore/scoreview.cpp]: add a couple spaces, between for and (, around the =  and after the ;

File: mscore/scoreview.cpp [hunk 5]
```diff
@@ -3474,13 +3593,15 @@ void ScoreView::dragNoteEntry(QMouseEvent* ev)
...
 void ScoreView::noteEntryButton(QMouseEvent* ev)
       {
-      QPointF p = toLogical(ev->pos());
-      _score->startCmd();
-      _score->putNote(p, ev->modifiers() & Qt::ShiftModifier);
-      _score->endCmd();
-      ChordRest* cr = _score->inputState().cr();
-      if (cr)
-            adjustCanvasPosition(cr, false);
+      if (ev->button()==Qt::LeftButton) {
```
Comment (Jojo-Schmitz) [inline mscore/scoreview.cpp]: spaces around the ==

**Comment by 19joho66** [comment]
Thanks for the hint JoJo. I've fixed your suggestions and sqashed everything to one commit and did a rebase. I hope everything is ok now.

And yes, you're right, the feature request is the one discussed in https://musescore.org/en/node/99536

**Comment by Jojo-Schmitz** [comment]
Looks much better now, just one space issue left ;-)

**Comment by Jojo-Schmitz** [comment]
Looks much better now, just one space issue left ;-)
And the commit message probably needs some fixing (git  commit --amend)
and removing .gitignore is not a good idea

**Comment by 19joho66** [comment]
Done...

**Comment by Jojo-Schmitz** [comment]
Well, the commit message still needs some work ;-)
The body text should probably be the title, and the body could be empty or have the URL to the discussion thread on musescore.org

**Comment by ZackTheCardshark** [comment]
@lasconic @19joho66 Notwithstanding the floated idea of reworking mouse entry to pop this up on a certain length of regular left-click, reception was generally positive for this one, as well. What about it?

**Comment by anatoly-os** [comment]
See https://github.com/musescore/MuseScore/pull/3609
